src/validation.cpp

```diff
@@ -2405,6 +2294,195 @@ bool BitcoindRPCCheck(const bool init)
     return true;
 }
 
+/* Takes federation redeeem script and adds HMAC_SHA256(pubkey, witnessProgram) as a tweak to each pubkey */
+CScript calculate_contract(const CScript& federationRedeemScript, const CScript& witnessProgram) {
+    CScript scriptDestination(federationRedeemScript);
+    int version;
+    std::vector<unsigned char> program;
+    if (!witnessProgram.IsWitnessProgram(version, program)) {
```

Comment:
Is this necessary? At some point bitcoin might switch to a new witness commitment format, e.g. when the sixteen remaining version bytes are exhausted or when quantum computers make the security of a 40 byte push untenable. Or for unforeseen reasons.

All you do is feed it into an HMAC later, so why does it care?